src/state/router.js

```diff
@@ -0,0 +1,142 @@
+
+import { REQUEST_TRIGGER, API_VERSIONS_RECEIVE, REQUEST_SELECT_ENDPOINT } from './actions';
+import { defaultState as defaultRequestState } from './request/reducer';
+import { defaultState as defaultUiState } from './ui/reducer';
+import { loadEndpoints } from './endpoints/actions';
+import { getEndpoints } from './endpoints/selectors';
+
+function getUrlParams() {
+	return new URL( window.location.href ).searchParams;
+}
+
+function encodeString( val ) {
+	return ( val === null || val === undefined ) ? '' : encodeURIComponent( val );
+}
+
+function encodeObject( val ) {
+	return ( typeof val === 'object' ) && Object.keys( val ).length === 0 ? '' : encodeURIComponent( JSON.stringify( val ) );
+}
+
+function decodeString( val ) {
+	return decodeURIComponent( val );
+}
+
+function decodeObject( val ) {
+	return JSON.parse( decodeURIComponent( val ) );
+}
+
+export const getUrlFromState = state => {
+
+	const { request, ui } = state;
+
+	const params = {
+		bodyParams: encodeObject( request.bodyParams ),
+		endpoint: encodeString( request.endpoint ? request.endpoint.pathLabeled : '' ),
+		method: encodeString( request.method ),
+		pathValues: encodeObject( request.pathValues ),
+		queryParams: encodeObject( request.queryParams ),
+		url: encodeString( request.url ),
+		api: encodeString( ui.api ),
+		version: encodeString( ui.version ),
+	};
+
+	// Discard empty params
+	Object.keys( params ).forEach( key => ! params[ key ] && ( delete params[ key ] ) );
```

Comment:
there's a value in this terseness, but we don't often use the pattern of `**` for control flow, and given how infrequently this project is maintained, it might be more valuable to choose something more familiar, including constructing the object positively instead of removing empty bits.

```js
const urlParams = new URLSearchParams();

for ( const [ value, key ] of Object.entries( params ) ) {
	if ( value ) {
		urlParams.update( key, value );
	}
}
```

also is the basic truthiness check valid here for all the parameters? it seems like we might end up needlessly encoding empty objects and omit `false` values.

I wonder if it would help to use `undefined` to include these when forming the `params` object. `url: '' === request.url ? encodeString( request.url ) : undefined`

Reply:
Nit: `const [ value, key ] of Object.entries( params )` should be `const [ key, value ] of Object.entries( params )`, or weird things start happening